mathics/core/atoms.py

```diff
@@ -640,14 +670,26 @@ def __new__(cls, real, imag):
         if isinstance(imag, MachineReal) and not isinstance(real, MachineReal):
             real = real.round()
 
-        self.real = real
-        self.imag = imag
+        value = (real, imag)
+        self = cls._complex_numbers.get(value)
+        if self is None:
 
-        # Set a value for self.__hash__() once so that every time
-        # it is used this is fast.
-        self.hash = hash(("Complex", real, imag))
+            self = super().__new__(cls)
+            self.real = real
+            self.imag = imag
+
+            self._value = value
+
+            # Cache object so we don't allocate again.
+            self._complex_numbers[value] = self
+
+            # Set a value for self.__hash__() once so that every time
+            # it is used this is fast. Note that in contrast to the
+            # cached object key, the hash key needs to be unique across all
+            # Python objects, so we include the class in the
+            # event that different objects have the same Python value
+            self.hash = hash((cls, value))
```

Comment:
Notice that the hash has a different form than the key. Is this intentional?

Reply:
Yes, again simplicity. While object hashes have to be unique across_all_ Mathics objects, keys for cache objects for a particular type, e.g. `String` or `Symbol `only need to be unique across that kind of object. And since the Mathics cache object represent a specific value in for that object, e.g. 5 for the Integer object, that value is be unique (for the Integer cache).

So taking String and Symbol as a good example (but note that right now we aren't caching strings for String), there could be a string called ``System`N`` and there _is_  a Symbol called  ``System`N``.  If we hashed these independent of the class these would come up with the same object hash. However in the type-specific object hash those values without the class name is unique. 

If one is debugging or working with the object hash, it is much nicer to work with the the value as its key rather than something more complicated.